Bitcoin (#283)
dfinity/portal: docs/developer-docs/functionality/bitcoin/bitcoin.md

-   Faster and cheaper transfers: ckBTC can be transferred with the low finality time of the Internet Computer (within seconds) and for a fraction of the cost of a Bitcoin transfer on the Bitcoin network. Using this scheme, only the settlement transfers with the Bitcoin network need to be done on the Bitcoin network, the majority of transfers could be done with lightning speed and low cost directly on the IC.

## How it Works -- Technology Background

Bitcoin-enabling the IC has required us to solve two advanced engineering challenges: (1) A protocol-level integration of the IC with the Bitcoin network and (2) a novel threshold ECDSA protocol.

**Protocol-level integration of the IC with the Bitcoin network**

Through the protocol-level integration of the IC with the Bitcoin network, the IC can obtain Bitcoin blocks and process the contained transactions. This allows for maintaining the full Bitcoin UTXO state on-chain on the IC and canisters can run queries against it. This allows canisters to know about the held UTXOs, and thus balance, of any Bitcoin address.

**Novel threshold ECDSA protocol**

> Canisters themselves can have ECDSA keys using a novel threshold ECDSA protocol, and so can receive and hold Bitcoin. Canisters can create Bitcoin transactions and submit them via the Bitcoin API to the Bitcoin network. They use the threshold ECDSA functionality to request threshold signatures on transactions to be submitted to the Bitcoin network. Threshold ECDSA is an extension of the IC\'s chain-key technology toolbox of protocols. Details regarding the IC\'s threshold ECDSA protocol can be found on the threshold ECDSA documentation page (link??).

Reviewer: Todo: check if apostrophes render correctly on both pages.

The protocol-level Bitcoin integration and threshold ECDSA protocol each expose an API on the management canister. Those APIs are the system-level APIs engineers use to write Bitcoin smart contracts on the IC.

> We next give a high-level overview of the abovementioned technology behind the direct Bitcoin integration. For details, we refer the reader to the Bitcoin page on the Internet Computer Wiki (link??) as well as the threshold ECDSA documentation page (link??).

Reviewer: ```suggestion
We next give a high-level overview of the abovementioned technology behind the direct Bitcoin integration. For details, we refer the reader to the Bitcoin page on the Internet Computer Wiki (link??) as well as the threshold ECDSA documentation page [here](./t-ecdsa.md).
```

### Protocol-level integration of the IC with the Bitcoin network

> We integrated the Internet Computer Protocol with the Bitcoin protocol to obtain a direct technical integration between the two networks. This integration can be activated per subnet of the Internet Computer. The integration serves two key purposes:

Reviewer: ```suggestion
We integrated the Internet Computer Protocol with the Bitcoin protocol to obtain a direct technical integration between the two networks. This integration can be activated on any number of Internet Computer subnets. At the beginning, there will only be one dedicated Bitcoin-activated subnet. The integration serves two key purposes:
```

-   Obtaining the Bitcoin UTXO set and keeping it on chain in the replicated state of the Internet Computer to be able to answer queries for UTXO sets and balances of Bitcoin accounts issued by canisters.

-   Accepting signed Bitcoin transactions of canisters and submitting them to the Bitcoin network.

> Figure

Reviewer: Is this a TODO?

Author: Yes!

Threshold ECDSA refers to the implementation of the ECDSA signature protocol using threshold cryptography. In a threshold ECDSA protocol, the private ECDSA key is secret shared between multiple parties and only an eligible quorum of the parties can generate a signature using their respective private key shares. The private key never exists in reconstructed form, but only in its secret-shared form. Key generation generates private key shares for the parties.

The threshold ECDSA protocol implemented as part of the Internet Computer\'s chain key technology toolbox uses a single master private key from which keys for canisters can be derived using BIP-32-like key derivation. Each canister has one such root key derived using the canister id and an arbitrary number of additional canister keys can be derived using a backward-compatible extension of BIP-32.

Threshold ECDSA on the Internet Computer will be rolled out to one signing subnet initially that will answer signing requests of canisters. The subnet enforces that only the canister that controls a key may request signatures with this key. All API calls go through Xnet traffic and thus incur some extra latency.

> Canisters can query their own or other canisters\' public keys, including further derived public keys of canisters. Canisters can request signatures with private keys they control, i.e., their root private key and derived private key. For requests of public keys or signatures with derived keys, a derivation path can be specified in the respective API.

Reviewer: ```suggestion
Canisters can query their own or other canisters\' public keys, including further derived public keys of canisters. Canisters can request signatures with private keys they control, i.e., their root private key and derived private key. For requests of public keys or signatures with derived keys, a derivation path can be specified in the respective API. You can find more details on threshold ECDSA [here](./t-ecdsa.md).
```

## Deployment Architecture

> The Bitcoin functionality will be activated on a single subnet of the IC and API calls from canisters to the Bitcoin API are routed via Xnet communication, thus extra latency is incurred. The feature may, if needed, in the future be additionally activated on (some) app subnets to avoid the additional Xnet latency and to scale out to be able to respond to more requests per time unit.

Reviewer: ```suggestion
The Bitcoin functionality will be activated on a single subnet of the IC and API calls from canisters to the Bitcoin API are routed via Xnet communication, thus extra latency is incurred. The feature may, if needed, in the future be additionally activated on (some) application subnets to avoid the additional Xnet latency and to be able to respond to more requests per time unit.
```

> Threshold ECDSA requests will equally be answered by a single active subnet, another subnet will back up the private key in secret-shared form for disaster recovery.

Reviewer: this feels like two separate sentences looking for a better home

## API

> The Bitcoin integration makes the following management canister APIs available to canisters (the threshold ECDSA API is explained in its documentation page and the interface specification). Each Bitcoin-related method needs to specify whether it uses Bitcoin \`mainnet\` or \`testnet\`.

Reviewer: ```suggestion
The Bitcoin integration makes the following management canister APIs available to canisters (the threshold ECDSA API is explained in [its documentation page](./t-ecdsa.md) and the [interface specification](../../../references/ic-interface-spec.md)). Each Bitcoin-related method needs to specify whether it uses Bitcoin \`mainnet\` or \`testnet\`.
```

> -   bitcoin_get_utxos: Given a get_utxos_request, which must specify a Bitcoin address and a Bitcoin network (mainnet or testnet), the function returns all unspent transaction outputs (UTXOs) associated with the provided address in the specified Bitcoin network based on the current view of the Bitcoin blockchain available to the Bitcoin component. The UTXOs are returned sorted by block height in descending order.\
>     > The optional filter parameter can be used to restrict the set of returned UTXOs, either providing a minimum number of confirmations or a page reference when pagination is used for addresses with many UTXOs. In the first case, only UTXOs with at least the provided number of confirmations are returned, i.e., transactions with fewer than this number of confirmations are not considered. In other words, if the number of confirmations is c, an output is returned if it occurred in a transaction with at least c confirmations and there is no transaction that spends the same output with at least c confirmations.\
>     > A get_utxos_request without the optional filter results in a request that considers the full blockchain, which is equivalent to setting min_confirmations to 0.
> 
> -   bitcoin_get_balance: Given a get_balance_request, which must specify a Bitcoin address and a Bitcoin network (mainnet or testnet), the function returns the current balance of this address in Satoshi (10\^8 Satoshi = 1 Bitcoin) in the specified Bitcoin network. The same address formats as for bitcoin_get_utxos are supported.
> 
> -   bitcoin_send_transaction: Given a send_transaction_request, which must specify a blob of a Bitcoin transaction and a Bitcoin network (mainnet or testnet), several checks are performed, and, if successful, the transaction is forwarded to the Bitcoin network.
> 
> -   bitcoin_get_current_fee_percentiles: The transaction fees in the Bitcoin network change dynamically based on the number of pending transactions. It must be possible for a canister to determine an adequate fee when creating a Bitcoin transaction.\
>     > This function returns the 100 fee percentiles, measured in millisatoshi/byte (10\^3 millisatoshi = 1 satoshi), over the last 10,000 transactions, i.e., over the transactions in the last approximately 4-10 blocks.

Reviewer: ```suggestion
-   `bitcoin_get_utxos`: Given a `get_utxos_request`, which must specify a Bitcoin address and a Bitcoin network (mainnet or testnet), the function returns all unspent transaction outputs (UTXOs) associated with the provided address in the specified Bitcoin network based on the current view of the Bitcoin blockchain available to the Bitcoin component. The UTXOs are returned sorted by block height in descending order.
    > The optional filter parameter can be used to restrict the set of returned UTXOs, either providing a minimum number of confirmations or a page reference when pagination is used for addresses with many UTXOs. In the first case, only UTXOs with at least the provided number of confirmations are returned, i.e. transactions with fewer than this number of confirmations are not considered. In other words, if the number of confirmations is c, an output is returned if it occurred in a transaction with at least c confirmations and there is no transaction that spends the same output with at least c confirmations.\
    > A `get_utxos_request` without the optional filter results in a request that considers the full blockchain, which is equivalent to setting `min_confirmations` to 0.

-   `bitcoin_get_balance`: Given a `get_balance_request`, which must specify a Bitcoin address and a Bitcoin network (mainnet or testnet), the function returns the current balance of this address in Satoshi (10\^8 Satoshi = 1 Bitcoin) in the specified Bitcoin network. The same address formats as for `bitcoin_get_utxos` are supported.

-   `bitcoin_send_transaction`: Given a `send_transaction_request`, which must specify a blob of a Bitcoin transaction and a Bitcoin network (mainnet or testnet), several checks are performed, and, if successful, the transaction is forwarded to the Bitcoin network.

-   `bitcoin_get_current_fee_percentiles`: The transaction fees in the Bitcoin network change dynamically based on the number of pending transactions. It must be possible for a canister to determine an adequate fee when creating a Bitcoin transaction.
    > This function returns the 100 fee percentiles, measured in millisatoshi/byte (10\^3 millisatoshi = 1 satoshi), over the last 10,000 transactions, i.e., over the transactions in the last approximately 4-10 blocks.
```
.md can be whacky with formatting. This is how my editor shows things correctly. todo: check on preview deployment

> We refer to the \[Internet Computer Interface Specification\](https://internetcomputer.org/docs/current/references/ic-interface-spec) for the details of the Bitcoin integration API.

Reviewer: ```suggestion
We refer to the [Internet Computer Interface Specification](../../../references/ic-interface-spec.md) for the details of the Bitcoin integration API.
```

-   The *SDK* for local development of canisters;

-   IC mainnet support as the pre-production environment for final testing on *Bitcoin testnet*;

-   IC mainnet support as the production environment for the release using *Bitcoin mainnet*.

### Local SDK

In the typical canister development workflow, canisters on the IC are compiled and run in the local environment using the Motoko or Rust canister SDK during their development. Thus, the SDK is the first stage of the development workflow. The SDK has been enabled to support both the Bitcoin integration and threshold ECDSA management canister APIs.

In contrast to the mainnet deployments of the feature, which integrate with Bitcoin testnet and Bitcoin mainnet, respectively, the SDK integrates with a locally-running bitcoind node in regression testing (regtest) mode. Using bitcoind in regtest mode is the preferred way of Bitcoin development. To facilitate our developers as best as possible, we integrated the SDK with bitcoind in regtest mode to bring the best Bitcoin development experience to the IC. Both development and automated testing of smart contracts are first done in the local environment.

> The Bitcoin adapter of the single replica running the local SDK environment connects to the local bitcoind node instead of multiple nodes of Bitcoin testnet or mainnet.

Reviewer: ```suggestion
The Bitcoin adapter of the single replica running the local SDK environment connects to the local bitcoind node instead of multiple nodes of Bitcoin testnet or mainnet. To see the relevant flags on dfx, please look at the output of `dfx start --help`.
```

### Bitcoin testnet on IC mainnet

Once a smart contract is ready for acceptance testing, it is deployed on IC mainnet (recall, there is no public IC testnet), still using the Bitcoin API set to connect to the Bitcoin testnet. This is the same setting used in local development. This setup is used to perform the acceptance testing of the dApp, using the Bitcoin testnet, i.e., no real value is at stake.

### Bitcoin mainnet on IC mainnet

The final stage of development of a Bitcoin smart contract is its deployment on IC mainnet with the Bitcoin API set to use Bitcoin mainnet. This is the final production environment for the smart contract. This is not yet available, but will be activated with the GA release.

## A Simple Example Canister -- Walkthrough

We will next show how to implement a simple example smart contract using the Bitcoin integration feature. Everything relevant will be explained, from installing the SDK, local development, to deploying the canister on IC mainnet and using the Bitcoin testnet integration.

> The following walkthrough uses a minimal example canister that uses all the APIs of the Bitcoin and threshold ECDSA functionality. This code is publicly available on GitHub (link) and can serve as the basis for your own Bitcoin integration project.

Reviewer: which link will this be on? I'm ok merging this even if the link is not quite live yet.

Our minimal example is a simple canister with the following functionality:

-   Expose a Bitcoin address derived from an ECDSA public key

-   Accept Bitcoin payments on this address (implicit)

-   Query UTXOs / balance for the address

-   Transfer Bitcoin to another Bitcoin address

Note that the focus here is on simplicity and therefore we will not go into all the details of how to handle aspects such as the pitfalls of UTXO handling in the light of transactions that do not get mined etc.

### Getting Started

> -   Download the latest SDK (minimum version ??)

Reviewer: ```suggestion
-   Download the latest SDK (minimum version 0.10.1)
```
AFAIK it's all implemented by now.

-   Create new project

> -   ...

Reviewer: I assume the rest is just an outline and will be filled later.
